### ISSUE TITLE
History processing function

### ISSUE THREAD

**Issue by JAForbes** [nushell/nushell#12276]
### Related problem

- Plain text history doesn't contain enough metadata to easily clean up after the fact
- History is duplicated which can lead to noisy searching
- Defining functions in the repl ends up in the history, but it is hard to navigate a multi line and complex expressions when trying to flip through your history with Up / Down
- There is no one true way to pruning history, it is largely a personal thing
- Moving history to sqlite solves the lack of context but cleaning up history is still a manual post process

### Describe the solution you'd like

It would be great to have a simple filter function that you can configure that receives each new command entered as text and then returns a boolean to state whether it should be included in the history.

From there a user could implement any special logic they'd like, and nushell could come with a base filter (e.g. skip commands prefixed with a space).

When there's a plugin ecosystem people could easily import a history module and this would be an important lifecycle hook for those plugins.

### Describe alternatives you've considered

I also considered a transform function that would allow you to alter the command.

Use cases would be e.g. grep for secrets in the command and replace them with variable names.  Transform relative paths to absolute paths for easier deduping etc.

I feel this could be added later but there is a possibility of corrupting history if there's a mistake in the logic and the filter approach is much safer.

**Comment by fdncred** [comment]
The sqlite history allows one to use `query db` and other commands to do many types of history processing. Not to mention, it's pretty easy to `open $nu.history-path | where ...` for filtering.

**Comment by JAForbes** [comment]
Yep that's great, but I think it would be great to have a pro-active process instead of a manual post process.

E.g. I'd rather define a function that filters out non zero exit code, any cd's / ls, all npm installs, anything referring to a secret etc.  It would be nice to define that function and it always works and I don't have to go and repair the history db as a daily/weekly chore.

I still really like the idea that history is in sqlite though.  I'd use that for removing things from history that are older than n weeks/months and haven't been used recently.  They are both useful.

**Comment by fdncred** [comment]
you can also do this type of thing via the nushell menu capability. For instance, here's one that was written to allow fzf usage.
```
  {
    name: fzf_menu_nu_ui
    only_buffer_difference: false
    marker: "# "
    type: {
      layout: list
      page_size: 10
    }
    style: {
      text: "#66ff66"
      selected_text: { fg: "#66ff66" attr: r }
      description_text: yellow
    }
    source: { |buffer, position|
      open $nu.history-path | get history.command_line
      | to text
      | fzf -f $buffer
      | lines
      | each { |v| {value: ($v | str trim) } }
    }
  }
```

**Comment by JAForbes** [comment]
Not sure what that is sorry.  But it looks pretty involved.  I'm thinking something as simple as

```nushell
def historyFilter [command: string] {
  if command | str trim != command { # etc
    false 
  } else { 
    true 
  }
}
$nu.history.filter = historyFilter
```

**Comment by KAAtheWiseGit** [comment]
I think this functionality fits in as a hook.  There are already hooks for prompts and environment, so adding one more for history wouldn't be that much of an issue, IMO.

**Comment by CrendKing** [comment]
Just to add a bit of legitimacy. PowerShell's ReadLine has an exactly the same feature [PSConsoleReadLineOptions.AddToHistoryHandler](https://learn.microsoft.com/en-us/dotnet/api/microsoft.powershell.psconsolereadlineoptions.addtohistoryhandler), which is a function that takes the command line string and return boolean of whether to add to history or not.

**Comment by jerrygreen** [comment]
I would love to skip all my «cd» commands from history. Being able to automatically filter what goes to history and what not, would be nice
